src/grpc/mod.rs

```diff
@@ -0,0 +1,197 @@
+
+use super::kbc_runtime;
+
+#[derive(Clone)]
+pub struct KeyProvider;
+
+impl KeyProviderService for KeyProvider {
+    fn un_wrap_key(
+        &mut self,
+        ctx: ::grpcio::RpcContext,
+        req: keyprovider::keyProviderKeyWrapProtocolInput,
+        sink: ::grpcio::UnarySink<keyprovider::keyProviderKeyWrapProtocolOutput>,
+    ) {
+        // Deserialize and parse the gRPC input to get KBC name, KBS uri and annotation.
+        let input_payload: InputPayload = match parse_input(req.KeyProviderKeyWrapProtocolInput) {
+            Ok(i) => i,
+            Err(e) => {
+                println!("Some error happend when parse the input: {}", e);
+                return ();
+            }
+        };
+
+        // Pass the KBC name, KBS uri and annotation to the KBC instance for content parsing and field decryption.
+        let kbc_runtime_mutex_clone = Arc::clone(&kbc_runtime::KBC_RUNTIME);
+        let mut kbc_runtime = match kbc_runtime_mutex_clone.lock() {
+            Ok(manager) => manager,
+            Err(e) => {
+                println!("Some error happend when access kbc manager(Mutex): {}", e);
+                return ();
+            }
+        };
+        let decrypted_optsdata = match kbc_runtime.decrypt(
+            input_payload.kbc_name,
+            input_payload.kbs_uri,
+            input_payload.annotation,
+        ) {
+            Ok(data) => data,
+            Err(e) => {
+                println!(
+                    "Some error happend when call kbc to decrypt optsdata: {}",
+                    e
+                );
+                return ();
+            }
+        };
+
+        // Construct output structure and serialize it as the return value of gRPC
+        let output_struct = KeyUnwrapOutput {
+            keyunwrapresults: KeyUnwrapResults {
+                optsdata: decrypted_optsdata,
+            },
+        };
+        let mut result = keyprovider::keyProviderKeyWrapProtocolOutput::new();
+        match serde_json::to_string(&output_struct) {
+            Ok(output_string) => {
+                result.KeyProviderKeyWrapProtocolOutput = output_string.as_bytes().to_vec();
+            }
+            Err(e) => {
+                println!("Some error happend when serialize the output struct: {}", e);
+                return ();
+            }
+        }
+        let output = sink
+            .success(result.clone())
+            .map_err(move |err| eprintln!("Failed to reply: {:?}", err))
+            .map(move |_| println!("Responded with result {{ {:?} }}", result));
+        ctx.spawn(output)
+    }
+
+    fn wrap_key(
+        &mut self,
+        ctx: ::grpcio::RpcContext,
+        req: keyprovider::keyProviderKeyWrapProtocolInput,
+        sink: ::grpcio::UnarySink<keyprovider::keyProviderKeyWrapProtocolOutput>,
+    ) {
+        let input_string = String::from_utf8(req.KeyProviderKeyWrapProtocolInput).unwrap();
+        println!("Request input {}", input_string);
```

Comment:
Is this going to print the unwrapped key to stdout? You will probably be running this in a trusted environment, but that still might not be ideal.